hwy/ops/arm_neon-inl.h

```diff
@@ -5770,7 +5829,41 @@ HWY_INLINE Vec128<T> MaxOfLanes(hwy::SizeTag<4> /* tag */, Vec128<T> v3210) {
   HWY_NEON_DEF_WIDE_PAIRWISE_REDUCTION(int16, 8, 4, name, prefix, s16)  \
   HWY_NEON_DEF_WIDE_PAIRWISE_REDUCTION(int8, 16, 8, name, prefix, s8)
 
+
+#define HWY_NEON_BUILD_RET_PAIRWISE_REDUCTION_VAL(type, size) Vec128<type##_t, size>
+#define HWY_NEON_DEF_PAIRWISE_REDUCTION_VAL(type, size, name, prefix, suffix)    \
+  HWY_API HWY_NEON_BUILD_RET_PAIRWISE_REDUCTION_VAL(type, size)                  \
+      name(hwy::SizeTag<sizeof(type##_t)>, Vec128<type##_t, size> v) {       \
+    HWY_NEON_BUILD_TYPE_T(type, size) tmp = prefix##_##suffix(v.raw, v.raw); \
+    if ((size / 2) > 1) tmp = prefix##_##suffix(tmp, tmp);                   \
+    if ((size / 4) > 1) tmp = prefix##_##suffix(tmp, tmp);                   \
+    return GetLane(tmp);            \
+  }
+#define HWY_NEON_DEF_WIDE_PAIRWISE_REDUCTION_VAL(type, size, half, name, prefix, \
+                                             suffix)                         \
+  HWY_API HWY_NEON_BUILD_RET_PAIRWISE_REDUCTION_VAL(type, size)                  \
```

Comment:
The thinking behind this was to avoid the additional vdup_lane_ followed by a vget_lane_ - though I may be missing a way to do this without the duplication of very similar macros

Reply:
Sure, I understand we want to avoid get(dup(x)). There are several strategies we can follow.

1) As pchx mentions, we can build the broadcasted-vector version out of dup('scalar' vector version). Here I see `HWY_NEON_BUILD_TYPE_T(type, size) tmp` which is a vector with only the first lane valid. So the SumOfLanes we are here defining could go into namespace detail, then the real SumOfLanes would be dup(detail::SumOfLanes) and SumOfLanesVal would be get(detail::SumOfLanes).

2) Where the current implement already naturally (for free) returns a broadcasted vector, e.g. wasm128 or x86_128, we can implement SumOfLanesVal as a generic GetLane(SumOfLanes()).

I think this avoids both duplication, and ever doing get(dup). Does that sound right?

Reply:
> Sure, I understand we want to avoid get(dup(x)). There are several strategies we can follow.
> 
> 1. As pchx mentions, we can build the broadcasted-vector version out of dup('scalar' vector version). Here I see `HWY_NEON_BUILD_TYPE_T(type, size) tmp` which is a vector with only the first lane valid. So the SumOfLanes we are here defining could go into namespace detail, then the real SumOfLanes would be dup(detail::SumOfLanes) and SumOfLanesVal would be get(detail::SumOfLanes).
> 2. Where the current implement already naturally (for free) returns a broadcasted vector, e.g. wasm128 or x86_128, we can implement SumOfLanesVal as a generic GetLane(SumOfLanes()).
> 
> I think this avoids both duplication, and ever doing get(dup). Does that sound right?

Thanks for explaining :)

For Armv7, a lot of the implementations were in the namespace detail - I've moved a bulk of these where the implementation results in a vector with the sum broadcast to all lanes outside this namespace. The only functions in the namespace detail now return the sum, which can then either be used by ReduceSum or SumOfLanes.

I think this has removed a lot of the duplication, but has resulted in a fair few of these functions being moved outside of the namespace detail (not quite sure if this is ok from a design perspective or not!)

Reply:
:) Sounds good!
No worries about moving functions out of detail::, that's just an implementation detail and I think it is OK to have more overloads in the top namespace instead of hiding them in detail::.